### ISSUE TITLE
Binary incompatiblilty for apps using library compiled against 3.0.3

### ISSUE THREAD

**Issue by dsyer** [reactor/reactor-core#339]
You can use the sample here to reproduce the problem: https://github.com/spring-cloud/spring-cloud-stream-samples/tree/master/reactive-processor-kafka. First run it as it is (with reactor-core 3.0.1) and it will launch successfully (and fail to connect to Kafka, but that's OK). Then change the reactor version to 3.0.4 and then the app fails to start:

```
Caused by: java.lang.NoSuchMethodError: reactor.core.publisher.Flux.subscribe(Ljava/util/function/Consumer;)Lreactor/core/Cancellation;
	at org.springframework.cloud.stream.reactive.FluxToMessageChannelResultAdapter.adapt(FluxToMessageChannelResultAdapter.java:47)
	at org.springframework.cloud.stream.reactive.FluxToMessageChannelResultAdapter.adapt(FluxToMessageChannelResultAdapter.java:33)
...
```

**Comment by yevhenii-melnyk** [comment]
The dependency for reactor-kafka should be 'io.projectreactor.kafka:reactor-kafka:1.0.0.BUILD-SNAPSHOT'.
The 3.0.4 reactor core dependency was added in 1.0.0.BUILD-SNAPSHOT version of reactor-kafka.

**Comment by mbogoevici** [comment]
@yevhenii-melnyk unfortunately, this is not depending on reactor-kafka in any way - this is simply about swapping version 3.0.1 of reactor with 3.0.4

**Comment by smaldini** [comment]
Yeah as explained offline, this is an oversight from us which I believe affect other libraries too. We simply tried to keep binary compatibility for subscribe by subtyping the still existing but deprecated `Cancellation` with a new `Disposable` type. User code was not expected to be affected by this with `subscribe(Xxx) : Disposable` change as `Disposable` is a `Cancellation`. This was especially an effort to prepare users for 3.1 changes, not to affect them in this way :(

Unfortunately that wasn't planned but we might need to carry on with this as changing again would break 3.0.4 now on maven-central, that means we need to add in our release notes https://github.com/reactor/reactor-core/releases/tag/v3.0.4.RELEASE this important issue and need to recompile project directly depending on reactor-core. I am very sorry about this, if anyone has a better idea we are open to suggestion but let us know.

**Comment by mbogoevici** [comment]
@smaldini So your suggestion is to upgrade everything to Reactor 3.0.4+ and consider any of the previous 3.0.x versions unreliable? Is this a strong commitment that everything in the 3.0.x world will be backwards compatible with 3.0.4?

**Comment by smaldini** [comment]
To give more visibility lets look forward at 3.0.5, possibly a 3.0.6 (not planned yet) and  3.1.

First the strongest argument in favor of stability is now we work with BOM starting from 3.0.4 (Aluminium-RELEASE).

- 3.0.5.RELEASE will be part of train release Aluminium-SR1 and will fix a few issues including the following API one: https://github.com/reactor/reactor-core/issues/338. It that might be fixed via a binary change but since it is unusable it won't affect anyone who at first cannot use it. 
Apart from this we just don't plan to affect any binary compatibility as it was expected for all 3.0.x unless exceptional reason like above or like the fact we accidentally introduced this very `subscribe(xxx)` issue. Timeline-wise we expect end of this month.

- 3.0.6.RELEASE will be part of train release Aluminium-SR2 if required, since we will be focusing more on 3.1.0.RELEASE and since we are meant to not break any API in patch release it should just be just fine, again if it will ever happen.

- 3.1.0.RELEASE will be part of train release Bismuth-RELEASE and will remove all deprecated plus a few edgy internal interfaces (but public). We expect this before Spring 5, so around March.

Note that all changes even the most minor API tweaks are referenced as in https://github.com/reactor/reactor-core/labels/API%20change and we just try to gather a maximum of those for 3.1 as we will be a long-term support dependency for spring 5 this time (tbh not that much was going to affect the way user would code beyond Disposable). These issues affect certainly library developers rather than end users and we of course try to limit the scope because Reactor is a reactive library geared toward application AND library developers.

**Comment by mbogoevici** [comment]
@smaldini 

From the Spring Cloud Stream perspective, I think we can even live with some breaking changes in 3.1. The situation ATM is that Stream (core) 1.1.1 and 1.2.0 will have to be rebuilt against 3.0.4 and are not binary compatible with apps based on 3.0.3. As a (carefully) documented one-off (including why pre-3.0.4 releases are not to be used), I don't see many other options, so that's our plan for now as per https://github.com/spring-cloud/spring-cloud-stream/issues/751. 

My comment was really out of the concern that we'd have to do a new Stream build for each 3.0.x release, which is not acceptable. To my understanding, the plans for 3.0.5 and 3.0.6 are such that this situation will not arise.

**Comment by smaldini** [comment]
Correct, its a regrettable uncaught exception. The only thing you can do is to recommend a more recent version of SCS to work alongside 3.0.4.+ to address that issue. Feeling like scala binary issue I know but this is exceptional.

**Comment by smaldini** [comment]
I'm going to mark this as closed for now and tagged with 3.0.4 if that's ok.
